spec/GraphQLOverHTTP.md

```diff
@@ -237,120 +311,304 @@ omitted.
+when using the `application/graphql+json` media type to encode the response
+body:
 
-The server SHOULD deny execution with a status code of `400` (Bad Request).
+##### Invalid request body or parsing failure
 
-Note: In certain circumstances, for example persisted queries that were
+For example: `NONSENSE`, `{"qeury": "{__typena`
+
+Requests that do not constitute a well-formed GraphQL request should result in
+status code `400` (Bad Request).
+
+##### Document validation failure
+
+Requests that fail to pass validation SHOULD be denied execution with a status
+code of `400` (Bad Request).
+
+Note: In certain circumstances, for example persisted operations that were
 previously known to be valid, the server MAY attempt execution regardless of
 validation errors.
 
-### Runtime validation
+Note: Validation rules include those specified in
+[the Validation section of the GraphQL specification](http://spec.graphql.org/draft/#sec-Validation),
+and any custom validation rules the server is using (for example: depth limit,
+complexity limit).
+
+##### Operation cannot be determined
+
+If [GetOperation()](<https://spec.graphql.org/draft/#GetOperation()>) raises a
+{GraphQL request error}, the server SHOULD NOT execute the request and SHOULD
+return a status code of `400` (Bad Request).
+
+##### Variable coercion failure
+
+If
+[CoerceVariableValues()](<https://spec.graphql.org/draft/#CoerceVariableValues()>)
+raises a {GraphQL request error}, the server SHOULD NOT execute the request and
+SHOULD return a status code of `400` (Bad Request).
+
+##### Field errors encountered during execution
+
+If the operation is executed and no {GraphQL request error} is raised then the
+server SHOULD respond with a status code of `200` (Okay). This is the case even
+if a {GraphQL field error} is raised during
+[GraphQL's ExecuteQuery()](<https://spec.graphql.org/draft/#ExecuteQuery()>) or
+[GraphQL's ExecuteMutation()](<https://spec.graphql.org/draft/#ExecuteMutation()>).
 
-Validation steps performed by resolvers during execution of the GraphQL
-operation.
+<!--
+When we add support for subscriptions,
+[GraphQL's MapSourceToResponseEvent()](<https://spec.graphql.org/draft/#MapSourceToResponseEvent()>)
+should be added to the above.
+-->
 
-The server SHOULD respond with a status code of `200` (Okay) to ensure clients
-receive a predictable result, no matter which fields they selected.
+Note: The GraphQL specification
+[differentiates field errors from request errors](https://spec.graphql.org/draft/#sec-Handling-Field-Errors)
+and refers to the situation wherein a {GraphQL field error} occurs as a partial
+response; it still indicates successful execution.
 
-### Client is not allowed to access the schema
+## Processing the response
 
-In case the client can not access the schema at all, the server SHOULD respond
-with the appropriate `4xx` status code.
+If the response uses a non-`200` status code and the media type of the response
```

Comment:
For what it's worth, the Apollo implementation of persisted queries does violate this MUST NOT: persisted query-related errors are sent (currently) as `application/json` with 400s and are parsed by the client.